libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
+        Node* left = UAVCAN_NULLPTR;
+        Node* right = UAVCAN_NULLPTR;
+        Node* equal_keys = UAVCAN_NULLPTR;
+    };
+
+    Node* root_;
+
+private:
+    size_t len_ ;
+
+    static int16_t heightOf(const Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return n->h;
+    }
+
+    void postOrderTraverseRecursively(Node* n, std::function<void(T*&)> forEach) {
+        if (n == UAVCAN_NULLPTR) {
+            return;
+        }
+
+        postOrderTraverseRecursively(n->left, forEach);
+        postOrderTraverseRecursively(n->right, forEach);
+        forEach(n->data);
+    }
+
+    Node* makeNode(T* payload) {
+        void* praw = this->allocator_.allocate(sizeof(Node));
+
+        if (praw == UAVCAN_NULLPTR) {
+            UAVCAN_TRACE("AvlTree", " OOM -- Can't allocate Node");
+            return UAVCAN_NULLPTR; // Push rejected
+        }
+
+        Node* node = new (praw) Node();
+        UAVCAN_ASSERT(node);
+
+        node->data = payload;
+        return node;
+    }
+
+    void deleteNode(Node*& n) {
+        if (n != UAVCAN_NULLPTR) {
+            n->~Node();
+            allocator_.deallocate(n);
+            n = UAVCAN_NULLPTR;
+        }
+    }
+
+    static int16_t balanceOf(Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return static_cast<int16_t>(heightOf(n->left) - heightOf(n->right));
```

Comment:
But if you got 65536 pending messages to transmit, you are going to run out of resources way sooner than filling up the tree. Keep in mind that expired nodes get removed too.

Reply:
> But if you got 65536 pending messages to transmit, you are going to run out of resources way sooner than filling up the tree.

That's a bold assumption that is too fragile to be as implicit as it is now (specifically, memory-rich applications like GNU/Linux-based nodes can theoretically exceed this limit). I think Scott is right.

Reply:
Aside from this part, the problem still exists on the int16_t chosen as the data type for each node's height.

I'm adding a static cast to the proposed changes so that the compiler doesn't treat the implicit warning as an error.

Reply:
> I think Scott is right..

I'll take what I can get at this point. Also, Scott is wrong about the static_cast. You do need it.